[JIT] remove list appends

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33297 [JIT] Allow mutated values as functional graph inputs
* **#33199 [JIT] remove list appends**
* #33186 [JIT] Pass To Safely Remove Aten Inplace Ops
* #33020 [JIT] Functional Graph Pass

Remove list appends when we can match them with a list construction. This helps create a larger functional graph

Differential Revision: [D20603187](https://our.internmc.facebook.com/intern/diff/D20603187)

Commit:
remove list appends

[ghstack-poisoned]

Commit:
Update on "[JIT] remove list appends"


Remove list appends when we can match them with a list construction. This helps create a larger functional graph

[ghstack-poisoned]

eellison commented:
@suo with these three PRs the basic block of resnet 18 from the `test_script_module_script_resnet` test gets the following graph after running inlining, constant propagation, remove mutation, and create functional graphs. The results are pretty good / what you would hope to see. Obviously resnet18 is not the be all end all but it's a good sign.


https://gist.github.com/eellison/3d46437dba6913bdb16d92203cad0042

Commit:
Update on "[JIT] remove list appends"


Remove list appends when we can match them with a list construction. This helps create a larger functional graph

[ghstack-poisoned]

Commit:
Update on "[JIT] remove list appends"


Remove list appends when we can match them with a list construction. This helps create a larger functional graph

[ghstack-poisoned]

Commit:
Update on "[JIT] remove list appends"


Remove list appends when we can match them with a list construction. This helps create a larger functional graph

[ghstack-poisoned]

Commit:
Update on "[JIT] remove list appends"


Remove list appends when we can match them with a list construction. This helps create a larger functional graph

Commit:
Update on "[JIT] remove list appends"


Remove list appends when we can match them with a list construction. This helps create a larger functional graph

suo reviewed (approved):
For the future: If we find ourselves adding more of these, they seem to follow the same structure:
1. find a pattern
2. try to make it atomic with the def
3. replace with "fused" thing

So we can factor out the common structure. But for just two patterns that seems overkill

suo commented on test/jit/test_functional_blocks.py:
```diff
@@ -126,3 +126,29 @@ def test_intermediary_use():
+
+    def test_remove_mutation_lists_append(self):
+        def successful_remove():
+            return [i for i in range(5)]
```
seems fine to suppress

Commit:
Update on "[JIT] remove list appends"


Remove list appends when we can match them with a list construction. This helps create a larger functional graph

Commit:
Update on "[JIT] remove list appends"


Remove list appends when we can match them with a list construction. This helps create a larger functional graph

Differential Revision: [D20603187](https://our.internmc.facebook.com/intern/diff/D20603187)
